File: Dynamic_Language_Python/src/Grid.py
"""
Grid module for Sudoku Solver.
"""
class Grid:
    """
    Grid class for representing a Sudoku grid.

    Attributes:
        GRID_SIZE (int): The size of the grid (9x9).
        BOX_SIZE (int): The size of each box (3x3).
        grid (list): A 2D list representing the Sudoku grid.

    Methods:
        __init__():
            Initialize the Grid object with default values for grid size, box size, and the grid itself.

        init(grid_str):
            Initialize the Grid object with a string representing the Sudoku grid.

        check_input(grid_str):
            Validate the input string for the Sudoku grid.

        parse_input(input_string):
            Parse the input string into a 2D grid (list of lists).

        get_row(row):
            Get the values in a specific row of the grid.

        get_column(col):
            Get the values in a specific column of the grid.

        get_box(row, col):
            Get the values in a specific 3x3 box of the grid.

        get_grid():
            Get the deep copy of current state of the grid.

        __iter__():
            Make the grid iterable.
    """
    def __init__(self):
        """
        Initialize the Grid object with default values for grid size, box size, and the grid itself.
        """
        self.GRID_SIZE = 9
        self.BOX_SIZE = 3
        self.grid = None

    def init(self, grid_str):
        """
        Initialize the Grid object with a string representing the Sudoku grid.

        Args:
            grid_str (str): A string representing the Sudoku grid, where each character is a digit.
        """
        self.parse_input(grid_str)

    def check_input(self, grid_str):
        """
        Validate the input string for the Sudoku grid.

        Args:
            grid_str (str): A string representing the Sudoku grid, where each character is a digit.

        Raises:
            ValueError: If the input string does not have the correct length.
            ValueError: If the input string contains non-digit characters.
            ValueError: If the input string contains out-of-range digits.
        """
        if len(grid_str) != self.GRID_SIZE ** 2:
            raise ValueError("Input string does not have the correct length.")
        if not all(char.isdigit() for char in grid_str):
            raise ValueError("Input string contains non-digit characters.")
        if not all(0 <= int(char) <= 9 for char in grid_str):
            raise ValueError("Input string contains out-of-range digits.")

    def parse_input(self, input_string):
        """
        Parse the input string into a 2D grid (list of lists).

        Args:
            input_string (str): A string representing the Sudoku grid, where each character is a digit.

        Raises:
            ValueError: If the input string does not have the correct length.
            ValueError: If the input string contains non-digit characters.
            ValueError: If the input string contains out-of-range digits.
        """
        self.check_input(input_string)
        self.grid = [[int(input_string[i * self.GRID_SIZE + j]) for j in range(self.GRID_SIZE)]
                     for i in range(self.GRID_SIZE)]

    def get_row(self, row):
        """
        Get the values in a specific row of the grid.

        Args:
            row (int): The index of the row to retrieve.

        Returns:
            list: A list of integers representing the values in the specified row.
        """
        return self.grid[row]

    def get_column(self, col):
        """
        Get the values in a specific column of the grid.

        Args:
            col (int): The index of the column to retrieve.

        Returns:
            list: A list of integers representing the values in the specified column.
        """
        return [self.grid[row][col] for row in range(self.GRID_SIZE)]

    def get_box(self, row, col):
        """
        Get the values in a specific 3x3 box of the grid.

        Args:
            row (int): The row index of a cell within the desired 3x3 box.
            col (int): The column index of a cell within the desired 3x3 box.

        Returns:
            list: A list of integers representing the values in the specified 3x3 box.
        """
        start_row = (row // self.BOX_SIZE) * self.BOX_SIZE
        start_col = (col // self.BOX_SIZE) * self.BOX_SIZE
        return [self.grid[r][c] for r in range(start_row, start_row + self.BOX_SIZE)
                for c in range(start_col, start_col + self.BOX_SIZE)]
    def get_grid(self):
        """
        Get the deep copy of current state of the grid.

        Returns:
            list: A list of lists representing the current state of the grid.
        """
        return [row[:] for row in self.grid]

    def __iter__(self):
        """
        Make the grid iterable.

        Returns:
            iterator: An iterator over the rows of the grid.
        """
        return iter(self.grid)

File: Static_Language_Java/src/main/java/Grid.java
/**
 * Grid class for Sudoku Solver in Java.
 */
public class Grid {

    final int GRID_SIZE = 9;
    final int BOX_SIZE = 3;
    int[][] grid;

    /**
     * Constructor for the Grid class.
     * Initializes the grid to null.
     */
    public Grid() {
        this.grid = null;
    }

    /**
     * Initializes the grid from a string input.
     *
     * @param gridStr the string representation of the grid
     * @throws IllegalArgumentException if the input string is invalid
     */
    public void init(String gridStr) throws IllegalArgumentException {
        parseInput(gridStr);
    }

    /**
     * Checks if the input string is valid.
     *
     * @param gridStr the string representation of the grid
     * @throws IllegalArgumentException if the input string does not have the correct length,
     *                                  contains non-digit characters, or contains out-of-range digits
     */
    public void checkInput(String gridStr) throws IllegalArgumentException {
        if (gridStr.length() != GRID_SIZE * GRID_SIZE) {
            throw new IllegalArgumentException("Input string does not have the correct length.");
        }
        for (char ch : gridStr.toCharArray()) {
            if (!Character.isDigit(ch)) {
                throw new IllegalArgumentException("Input string contains non-digit characters.");
            }
            int digit = Character.getNumericValue(ch);
            if (digit < 0 || digit > 9) {
                throw new IllegalArgumentException("Input string contains out-of-range digits.");
            }
        }
    }

    /**
     * Parses the input string to a 2D grid.
     *
     * @param inputString the string representation of the grid
     * @throws IllegalArgumentException if the input string is invalid
     */
    public void parseInput(String inputString) throws IllegalArgumentException {
        checkInput(inputString);
        grid = new int[GRID_SIZE][GRID_SIZE];
        for (int i = 0; i < GRID_SIZE; i++) {
            for (int j = 0; j < GRID_SIZE; j++) {
                grid[i][j] = Character.getNumericValue(inputString.charAt(i * GRID_SIZE + j));
            }
        }
    }

    /**
     * Gets a specific row from the grid.
     *
     * @param row the index of the row to retrieve
     * @return an array representing the row
     */
    public int[] getRow(int row) {
        return grid[row];
    }

    /**
     * Gets a specific column from the grid.
     *
     * @param col the index of the column to retrieve
     * @return an array representing the column
     */
    public int[] getColumn(int col) {
        int[] column = new int[GRID_SIZE];
        for (int row = 0; row < GRID_SIZE; row++) {
            column[row] = grid[row][col];
        }
        return column;
    }

    /**
     * Gets a specific box (3x3 sub-grid) from the grid.
     *
     * @param row the row index of the top-left cell of the box
     * @param col the column index of the top-left cell of the box
     * @return an array representing the box
     */
    public int[] getBox(int row, int col) {
        int[] box = new int[BOX_SIZE * BOX_SIZE];
        int startRow = (row / BOX_SIZE) * BOX_SIZE;
        int startCol = (col / BOX_SIZE) * BOX_SIZE;
        int index = 0;
        for (int r = startRow; r < startRow + BOX_SIZE; r++) {
            for (int c = startCol; c < startCol + BOX_SIZE; c++) {
                box[index++] = grid[r][c];
            }
        }
        return box;
    }

    /**
     * Gets a copy of the grid.
     *
     * @return a 2D array representing the grid
     */
    public int[][] getGrid() {
        int[][] gridCopy = new int[GRID_SIZE][GRID_SIZE];
        for (int i = 0; i < GRID_SIZE; i++) {
            gridCopy[i] = grid[i].clone();
        }
        return gridCopy;
    }

    /**
     * Gets an iterator to iterate over the grid.
     *
     * @return a 2D array representing the grid
     */
    public int[][] getGridIterator() {
        return grid;
    }
}